packages/@aws-cdk/aws-servicecatalog/lib/portfolio.ts

```diff
@@ -0,0 +1,246 @@
+    const hashId = this.getKeyForPortfolio(identifier);
+    new CfnPortfolioPrincipalAssociation(this, `PortolioPrincipalAssociation${hashId}`, {
+      portfolioId: this.portfolioId,
+      principalArn: principalArn,
+      principalType: PrincipalType.IAM,
+    });
+  }
+
+  /**
+   * Create a unique id based off the L1 CfnPortfolio or the arn of an imported portfolio.
+   */
+  protected abstract getKeyForPortfolio(value: string): string;
+}
+
+/**
+ * Properties for a Portfolio.
+ */
+export interface PortfolioProps {
+
+  /**
+   * The name of the portfolio.
+   */
+  readonly name: string;
+
+  /**
+   * The provider name.
+   */
+  readonly provider: string;
+
+  /**
+   * The accept language.
+   * @default - No accept language provided
+   */
+  readonly acceptLanguage?: AcceptLanguage;
+
+  /**
+   * Description for portfolio.
+   * @default - No description provided
+   */
+  readonly description?: string;
+}
+
+/**
+ * A Service Catalog portfolio.
+ */
+export class Portfolio extends PortfolioBase {
+
+  /**
+   * Creates a Portfolio construct that represents an external portfolio.
+   *
+   * @param scope The parent creating construct (usually `this`).
+   * @param id The construct's name.
+   * @param portfolioArn the Amazon Resource Name of the existing portfolio.
+   */
+  public static fromPortfolioArn(scope: Construct, id: string, portfolioArn: string): IPortfolio {
+    const arn = Stack.of(scope).parseArn(portfolioArn);
+    const portfolioId = arn.resourceName;
+
+    if (!portfolioId) {
+      throw new Error('Missing required Portfolio ID from Portfolio ARN: ' + portfolioArn);
+    }
+
+    class Import extends PortfolioBase {
+      portfolioArn = portfolioArn;
+      portfolioId = portfolioId!;
+
+      protected getKeyForPortfolio(value: string): string {
+        return getIdentifier(this.portfolioArn, value);
+      }
+    }
+    return new Import(scope, id, {
+      environmentFromArn: portfolioArn,
+    });
+  }
+
+  public readonly portfolioArn: string;
+  public readonly portfolioId: string;
+  private readonly portfolio: CfnPortfolio;
+
+  constructor(scope: Construct, id: string, props: PortfolioProps) {
+    super(scope, id);
+
+    this.validatePortfolioProps(props, id);
+
+    this.portfolio = new CfnPortfolio(this, 'Resource', {
+      displayName: props.name,
+      providerName: props.provider,
+      description: props.description,
+      acceptLanguage: props.acceptLanguage,
+    });
+
+    this.portfolioId = this.portfolio.ref;
+
+    this.portfolioArn = Stack.of(this).formatArn({
+      service: 'servicecatalog',
+      resource: 'portfolio',
+      resourceName: this.portfolioId,
+    });
+  }
+
+  protected getKeyForPortfolio(value: string): string {
+    return getIdentifier(Names.nodeUniqueId(this.portfolio.node), value);
+  }
+
+  private validatePortfolioProps(props: PortfolioProps, id: string) {
+    InputValidator.validateLength(id, 'portfolio name', 1, 100, props.name);
```

Comment:
I don't think I would include the `id` of the resource in the error message - it's not unique in the CDK app.

If you really want to pass something like that, pass `this.node.path`.